### PR TITLE
WT-13080 Pause background compaction in case of cache pressure

### PR DESCRIPTION
**Pull request by etienneptl**
Make the background compaction thread sleep if the cache has reached `eviction_trigger` or if dirty eviction is needed.

### CODE REVIEW

File: src/conn/conn_compact.c [hunk 1]
```diff
@@ -602,6 +620,11 @@ __background_compact_server(void *arg)
         if (!running)
             continue;
 
+        /* Check if there is cache pressure. */
+        cache_pressure = __wt_cache_bytes_inuse(cache) >= cache_pressure_limit;
```
Comment (quchenhao) [inline src/conn/conn_compact.c:624]: Should we consider dirty eviction threshold and update threshold as well?

Comment (etienneptl) [reply src/conn/conn_compact.c:624]: We tried this and unfortunately, it's very easy to hit those thresholds while the system is not in trouble and eviction is doing its job. We thought that checking if we have a cache almost full would be the best way to detect that "the system might be in trouble soon"

Comment (quchenhao) [reply src/conn/conn_compact.c:624]: You mean it is easy to reach dirty and update trigger?

Comment (etienneptl) [reply src/conn/conn_compact.c:624]: Yes

File: src/conn/conn_compact.c [hunk 2]
```diff
@@ -549,8 +553,13 @@ __background_compact_server(void *arg)
             __wt_spin_unlock(session, &conn->background_compact.lock);
         }
 
-        /* When the entire metadata file has been parsed, take a break or wait until signalled. */
-        if (full_iteration || !running) {
+        /*
+         * Take a break or wait until signalled in the following conditions:
```
Comment (tod-johnson-mongodb) [inline src/conn/conn_compact.c:557]: Change "in the following conditions" to "in any of the following conditions"

File: src/docs/arch-compact.dox [hunk 3]
```diff
@@ -81,7 +81,8 @@ Background compaction retrieves the files to compact by walking the metadata fil
 background compaction reduces the number of checkpoints by skipping the first one described in the
 section above. Furthermore, background compaction gathers statistics over time to keep track of tables
 where compaction is successful. Using those statistics, the service focuses on tables where compaction
-is likely to reclaim space and avoids unnecessary attempts.
+is likely to reclaim space and avoids unnecessary attempts. Finally, if \c eviction_dirty_trigger has been reached,
```
Comment (tod-johnson-mongodb) [inline src/docs/arch-compact.dox:84]: This doesn't mention the second reason why background compaction pauses: the cache is almost full.

File: src/conn/conn_compact.c [hunk 4]
```diff
@@ -518,19 +518,23 @@ __background_compact_find_next_uri(WT_SESSION_IMPL *session, WT_ITEM *uri, WT_IT
...
     WT_SESSION *wt_session;
     WT_SESSION_IMPL *session;
-    bool full_iteration, running;
+    double cache_pressure_limit;
+    bool cache_pressure, full_iteration, running;
 
     session = arg;
     conn = S2C(session);
+    cache = conn->cache;
+    cache_pressure_limit = WT_EVICT_PRESSURE_THRESHOLD * conn->cache_size;
```
Comment (tod-johnson-mongodb) [inline src/conn/conn_compact.c:535]: Why is WT_EVICT_PRESSURE_THRESHOLD a constant?

File: src/conn/conn_compact.c [hunk 5]
```diff
@@ -602,6 +616,17 @@ __background_compact_server(void *arg)
         if (!running)
             continue;
 
+        /*
+         * Throttle background compaction if one of the following conditions is met:
```
Comment (tod-johnson-mongodb) [inline src/conn/conn_compact.c:620]: Change "if one or" to "if any of"

**Review by tod-johnson-mongodb — CHANGES_REQUESTED** [review]
LGTM except for one thing: Does "if the cache size has not yet been set in a shared cache" need to be handled?  See my comment on conn_compact.c new line 535.

All other comments are optional suggestions.

**Review by tod-johnson-mongodb — APPROVED** [review]
I can't enter "Request changes" since I am on vacation next week and unable to respond to you. Thus I have changed this to Approve. However the concern about "if the cache size has not yet been set in a shared cache" still needs to be handled or proven OK.

File: src/conn/conn_compact.c (hunk 4 above)
Comment (etienneptl) [reply src/conn/conn_compact.c:535]: We are checking two things with background compaction:

- The cache is not almost full, hence the usage of `WT_EVICT_PRESSURE_THRESHOLD * conn->cache_size`. I am happy to add `+ 1` here to check for shared cache.
- Dirty eviction is not needed by checking `__wt_eviction_dirty_needed()`

File: src/conn/conn_compact.c (hunk 2 above)
Comment (etienneptl) [reply src/conn/conn_compact.c]: Thank you, see 8b097254fff78dae84013b9d4846d84f31e4423b

File: src/docs/arch-compact.dox (hunk 3 above)
Comment (etienneptl) [reply src/docs/arch-compact.dox]: I missed it, see 8b097254fff78dae84013b9d4846d84f31e4423b.

File: src/conn/conn_compact.c (hunk 5 above)
Comment (etienneptl) [reply src/conn/conn_compact.c]: Fixed in 42a90cb0823fd76c2ce939d911bd9be8839ae8ed

File: src/conn/conn_compact.c (hunk 4 above)
Comment (etienneptl) [reply src/conn/conn_compact.c:535]: If the cache size is not set yet, `cache_pressure_limit` would be 0 which does not make much sense.
I could set `cache_pressure_limit` inside the loop where it's used to make sure it gets set at some point.
But @Siddhartha8899, is that a thing that with shared cache we may have a value of 0 for `conn->cache_size`?

File: src/conn/conn_compact.c [hunk 6]
```diff
@@ -602,6 +616,17 @@ __background_compact_server(void *arg)
         if (!running)
             continue;
 
+        /*
+         * Throttle background compaction if any of the following conditions is met:
+         * - The dirty trigger threshold has been reached as compaction may generate more dirty
+         * content.
```
Comment (quchenhao) [inline src/conn/conn_compact.c:622]: Can we add a comment about why we don't consider update threshold here? We only mark the page dirty but not adding any new updates.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/conn/conn_compact.c (hunk 6 above)
Comment (etienneptl) [reply src/conn/conn_compact.c]: Sure, added in 3f3320e6e1da3084d3c10e3eb9c48f44cd1aca65.

File: src/conn/conn_compact.c (hunk 4 above)
Comment (etienneptl) [reply src/conn/conn_compact.c]: After discussing with @Siddhartha8899, he suggested using `__wt_eviction_clean_needed()` as it is an existing function that checks if the cache content is greater than `eviction_trigger`. This is a better solution than using `WT_EVICT_PRESSURE_THRESHOLD` as both `WT_EVICT_PRESSURE_THRESHOLD`  and `eviction_trigger` have a value of 95%. Furthermore, relying on `eviction_trigger` is better for the application as it's a configurable parameter.
Thank you, Sid!
See 09a4b1f7c669f7a314facfc12ba5a3da47a8aca5

**Review by mm-ng — APPROVED** [review]
LGTM, I'm happy with this approach.

File: src/conn/conn_compact.c (hunk 4 above)
Comment (Sean04) [reply src/conn/conn_compact.c]: Should prefetch also use `__wt_eviction_clean_needed` then?

Comment (etienneptl) [reply src/conn/conn_compact.c]: Yes, @Siddhartha8899 and I made the same observation.
cc @mm-ng

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Review by Sean04 — APPROVED** [review]
LGTM

**Pull request merged**